README.md
# pystockwatch
This toy package is created as part of a group project for DSCI524 of UBC-MDS program. Main idea of the package is to provide a very simple interface for checking movements in stock prices in comparison to the market. This will be implemented by accessing public data from Yahoo Finance by selecting a ticker of the stock, range of start and end dates and benchmark for comparison, such as SP500 or NASDAQ. As a result of data processing with pystockwatch package, user will be able to see a plot with two lines showing percentage change since the start date and volumes of trading. Example of such output from commercial software "TradingView.com" is given below, which will be replaced later with actual example from pystockwatch output

 <img src="docs/example.PNG">
 
 #### Functions description
 
 This package will be built with following four functions:
 - `percent_change`
 Calculation of profit percentage change of a stock for a given period of time based on data extracted from Yahoo finance.
 - `profit_viz`
 Visualization of profit percentage change trend of a stock against market benchmark for a given period of time.  
 - `volume_change`
 Calculation of daily trading volume change of a stock, whether it is increasing or decreasing. This data will be used in then next function for color mapping of bar plot.
 - `volume_viz`
 Visualization of trading volume as bar plots colored by changes in volume in comparison with previous day. This plot will be overlayed with line plots created with the second function.
  
 More information about those functions can be found in docstrings of `/src/pystockwatch/pystockwatch.py`
 
#### Comparison with similar packages
There are many packages written for analysis of stock data. One of the most popular Python package in this category is `pyti`, which gives a lot of manipulations with time series data, such as creating moving averages of stock price or calculating hundreds of parameters for technical analysis. Power of `pystockwatch` is in its simplicity, so users are not overwhelmed with all extra features and just have a simple view with key comparisons for a stock of interest

## Installation

```bash
$ pip install pystockwatch
```

## Usage
Usage clarification will be updated after completion of all functions. In general, after installation of this package user will be required to input four parameters: `stock_ticker`, `start_date`, `end_date` and `benchmark_ticker`. Output results will be in form of an interactive plot based on "plotly" package. 

## Contributors

Affrin Sultana, Helin Wang, Pavel Levchenko, Shi Yan Wang

## Dependencies
To use this package please use Python version 3.8 or higher. Also download several dependencies with [yaml file](https://github.com/UBC-MDS/pystockwatch/blob/main/dependencies-pystockwatch.yaml) by running the following command at the command line/terminal
``` bash
conda env create --file dependencies-pystockwatch.yaml
conda activate pystockwatch
```

## License

`pystockwatch` was created by Affrin Sultana, Helin Wang, Pavel Levchenko, Shi Yan Wang. It is licensed under the terms of the MIT license.

## Credits

`pystockwatch` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (13/01/2022)

- First release of `pystockwatch`!

tests/test_pystockwatch.py
from pystockwatch import pystockwatch
from pytest import raises

def test_percent_change():
    # Invalid input for stock ticker
    with raises(NameError) as error_ticker:
        percent_change("Some Stock", "2017-01-01", "2017-01-10")
    assert(
        str(error_ticker.value) == "You enter an invalid stock ticker! Try again.")

    # Invalid input for start date
    with raises(ValueError) as error_start_date:
        percent_change("AAPL", "20170101", "2017-01-10")
    assert(
        str(error_start_date.value) == "You enter an invalid start date! Try date formatted in YYYY-MM-DD.")

    # Invalid input for end date
    with raises(ValueError) as error_end_date:
        percent_change("AAPL", "2017-01-01", "20170110")
    assert(
        str(error_end_date.value) == "You enter an invalid end date! Try date formatted in YYYY-MM-DD.")

    # End date is ealier than start date
    with raises(ValueError) as error_end_date:
        percent_change("AAPL", "2017-01-10", "2017-01-01")
    assert(
        str(error_end_date.value) == "End date should not be earlier than start date! Try again.")


def test_profit_viz():
  pass


def test_volume_change():
    # Invalid stock ticker
    test_1 = "invalid ticker"
    with raises(NameError) as error_ticker:
        volume_change(test_1, "2021-12-01", "2021-12-31")
    assert(
        str(error_ticker.value) == "You enter an invalid stock ticker! Try again.")
    
    # Invalid date
    test_2 = "invalid start_date"
    with raises(ValueError) as error_start_date:
        volume_change("A", "2021", "2021-01-01")
    assert(
        str(error_start_date.value) == "You enter an invalid start date! Try again.")
    
    test_3 = "invalid_end_date"
    with raises(ValueError) as error_end_date:
        volume_change("A", "2021-01-01", 100)
    assert(
        str(error_end_date.value) == "You enter an invalid end date! Try again.")

    # Test indicators values
    indicators = ['nan', 'Decrease', 'Increase']
    df = volume_change("A", "2017-01-01", "2017-01-01")
    df_unique = df["Volume_Change"].unique()
    assert set(df_unique) <= set(indicators)
    

def volume_viz():
  pass


tests/test_percent_change.py
from pystockwatch.percent_change import percent_change
from pytest import raises

def test_percent_change():
    # Invalid input for stock ticker
    with raises(NameError) as error_ticker:
        percent_change("Some Stock", "2017-01-01", "2017-01-10")
    assert(
        str(error_ticker.value) == "You enter an invalid stock ticker! Try again.")

    # Invalid input for start date
    with raises(ValueError) as error_start_date:
        percent_change("AAPL", "20170101", "2017-01-10")
    assert(
        str(error_start_date.value) == "You enter an invalid start date! Try date formatted in YYYY-MM-DD.")

    # Invalid input for end date
    with raises(ValueError) as error_end_date:
        percent_change("AAPL", "2017-01-01", "20170110")
    assert(
        str(error_end_date.value) == "You enter an invalid end date! Try date formatted in YYYY-MM-DD.")

    # End date is ealier than start date
    with raises(ValueError) as error_end_date:
        percent_change("AAPL", "2017-01-10", "2017-01-01")
    assert(
        str(error_end_date.value) == "End date should not be earlier than start date! Try again.")

src/pystockwatch/percent_change.py
import pandas as pd
import numpy as np
import yfinance as yf


def percent_change(stock_ticker, start_date, end_date):
    """
    Calculates daily percentage change of a stock price within a given period of time
    
    Parameters
    ----------
    stock_ticker : string 
        Ticker of the stock such as 'AAPL', or 'AAPL MSFT SPY' for multiple tickers
    start_date : string
        Initial date for data extraction
    end_date : string
        Final date for stock analysis
    
    Returns
    --------
    DataFrame
        A data frame with dates and their corresponding stock price percentage changes.
    
    Examples
    --------
    >>> percent_change('AAPL', '2017-01-01', '2017-01-10')
                    Price Change Percentage(%) 
              Date
        2017-01-03                      0.0000
        2017-01-04                     -0.1119
        2017-01-05                      0.3960
        2017-01-06                      1.5153
        2017-01-09                      2.4451
    >>> percent_change('AAPL MSFT', '2017-01-01', '2017-01-10')
                    Price Change Percentage(%) 
                              AAPL        MSFT
              Date
        2017-01-03          0.0000      0.0000
        2017-01-04         -0.1119     -0.4474
        2017-01-05          0.3960     -0.4474
        2017-01-06          1.5153      0.4155
        2017-01-09          2.4451      0.0959
    """ 
    
    # Import original dataframe by giving stock ticker, start data and end date
    data = yf.download(stock_ticker, start=start_date, end=end_date)
    
    # Only Keep "Adj Close" Price for 
    data = data.drop(columns={'Open', 'High', 'Low', 'Close', 'Volume'})
    
    # Carry out calculation
    for i in range(1,len(data)):
        data.iloc[i,:] = (data.iloc[i,:] - data.iloc[0,:])/data.iloc[0,:]*100
    
    data.iloc[0,:] = (data.iloc[0,:] - data.iloc[0,:])/data.iloc[0,:]*100
    
    # Manipulate column name
    data = data.rename(columns={"Adj Close": "Price Change Percentage(%)"})
    
    # Return result
    return pd.DataFrame(data)

src/pystockwatch/pystockwatch.py
# authors: Affrin Sultana, Helin Wang, Shi Yan Wang and Pavel Levchenko

import plotly.express as px
import plotly.graph_objects as go
import pandas as pd
import numpy as np
import yfinance as yf
import pandas_datareader as pdr
import datetime
import warnings

def percent_change(stock_ticker, start_date, end_date):
    """
    Calculates daily percentage change of a stock price within a given period of time
    
    Parameters
    ----------
    stock_ticker : string 
        Ticker of the stock such as 'AAPL', or 'AAPL MSFT SPY' for multiple tickers
    start_date : string
        Initial date for data extraction
    end_date : string
        Final date for stock analysis
    
    Returns
    --------
    DataFrame
        A data frame with dates and their corresponding stock price percentage changes.
    
    Examples
    --------
    >>> percent_change('AAPL', '2017-01-01', '2017-01-10')
                    Price Change Percentage(%) 
              Date
        2017-01-03                      0.0000
        2017-01-04                     -0.1119
        2017-01-05                      0.3960
        2017-01-06                      1.5153
        2017-01-09                      2.4451
    >>> percent_change('AAPL MSFT', '2017-01-01', '2017-01-10')
                    Price Change Percentage(%) 
                              AAPL        MSFT
              Date
        2017-01-03          0.0000      0.0000
        2017-01-04         -0.1119     -0.4474
        2017-01-05          0.3960     -0.4474
        2017-01-06          1.5153      0.4155
        2017-01-09          2.4451      0.0959
    """ 
    
    # Assert ticker input value
    ticker = yf.Ticker(stock_ticker)
    if(ticker.info["regularMarketPrice"] == None):
        raise NameError("You enter an invalid stock ticker! Try again.")
    
    # Assert start date input value
    format = "%Y-%m-%d"
    try: datetime.datetime.strptime(start_date, format)
    except ValueError:
        raise ValueError("You enter an invalid start date! Try date formatted in YYYY-MM-DD.")
    
    # Assert end date input value
    try: datetime.datetime.strptime(end_date, format)
    except ValueError:
        raise ValueError("You enter an invalid end end! Try date formatted in YYYY-MM-DD.")

    # Assert end date is later than start date
    format = "%Y-%m-%d"
    if(datetime.datetime.strptime(end_date, format) < datetime.datetime.strptime(start_date, format)):
        raise ValueError("You enter an end date which is earlier than the start date! Try again.")
    
    # Import original dataframe by giving stock ticker, start data and end date
    data = yf.download(stock_ticker, start=start_date, end=end_date)
    
    # Only Keep "Adj Close" Price for 
    data = data.drop(columns={'Open', 'High', 'Low', 'Adj Close', 'Volume'})
    
    # Carry out calculation
    for i in range(1,len(data)):
        data.iloc[i,:] = (data.iloc[i,:] - data.iloc[0,:])/data.iloc[0,:]*100
    
    data.iloc[0,:] = (data.iloc[0,:] - data.iloc[0,:])/data.iloc[0,:]*100
    
    # Manipulate column name
    data = data.rename(columns={"Close": "Price Change Percentage(%)"})
    
    # Return result
    return pd.DataFrame(data)


def profit_viz(stock_ticker, start_date, end_date, benchmark_ticker):
    """
    Visualizes trend of a stock price change against the market benchmark within a given period of time
    
    Parameters
    ----------
    stock_ticker : string
        Ticker of the stock such as 'AAPL'
    start_date : string 
        Initial date for data extraction
    end_date : string
        Final date for stock analysis
    benchmark_ticker : string 
        Ticker for benchmark comparison such as 'SPX' 
    
    Returns
    --------
    Interactive line plots which shows percentage change in stock price and market performance over time 
    
    Examples
    --------
    >>> profit_viz('AAPL', '01-01-2015', '01-01-2022', 'SPX')
    """
    pass
    # TODO

    
def volume_change(stock_ticker, start_date, end_date):
        """ 
    Calculates the daily trading volume change status of a stock within a given period of time
    Parameters
    ----------
    stock_ticker : string 
        Ticker of the stock such as 'AAPL'
    start_date : string
        Initial date for data extraction
    end_date : string 
        Final date for stock analysis
        
    Returns
    --------
        A data frame with dates and their corresponding trading volume changes
    Examples
    --------
        >>> volume_change('AAPL', '2021-01-01', '2022-01-01')
        Date             Volume Volume_Change
        01-01-2021        1000        Nan
        01-02-2021        2000        Increase
        01-03-2021        3000        Increase
        01-04-2021        2500        Decrease
        ...
        12-31-2021        4000        Increase
        01-01-2022        5000        Increase
    """
    yf.pdr_override()
    # Assert ticker value
    ticker = yf.Ticker(stock_ticker)
    if(ticker.info["regularMarketPrice"] == None):
        raise NameError("You enter an invalid stock ticker! Try again.")
    # Assert date value
    format = "%Y-%m-%d"
    try: datetime.datetime.strptime(start_date, format)
    except ValueError:
        raise ValueError("You enter an invalid start date! Try again.")
    try: datetime.datetime.strptime(end_date, format)
    except ValueError:
        raise ValueError("You enter an invalid end end! Try again.")
    df = pdr.get_data_yahoo(stock_ticker, start=start_date, end=end_date)['Volume'].reset_index()
    # Assert correct data fetched
    try:
        isinstance(df, pd.DataFrame)
    except ValueError:
        raise ValueError("You input can't be converted to a pandas dataframe.")
    df['Volume_dif'] = df['Volume'].diff().to_frame()
    df["Volume_Change"] = np.select([df["Volume_dif"] > 0, df["Volume_dif"]<0], ["Increase", "Decrease"], 
                                    default = np.nan)
    # Assert correct indicator values
    for indicator in df["Volume_Change"]:
        if(indicator != "Decrease" and indicator != "Increase" and indicator != "nan"):
            raise ValueError("Incorrect Volume Change indicator")
    return df[['Date','Volume', 'Volume_Change']]
    

def volume_viz(stock_ticker, start_date, end_date):
    """
    Visualize the daily trading volume of a stock using bar plot within a given period of time

    Parameters
    ----------
    stock_ticker : string  
        Ticker of the stock such as 'AAPL'
    start_date : string 
        Initial date for data extraction
    end_date : string 
        Final date for stock analysis
    
    Returns
    --------
    Interactive plot with overlay of line plots and bar plot
    
    Examples
    --------
    >>> volume_viz('AAPL', '01-01-2015', '01-01-2022')
    """
    vdf = volume_change(stock_ticker, start_date, end_date)
    vdf_increase = vdf.loc[vdf['Volume_Change']=='Increase']
    vdf_decrease = vdf.loc[vdf['Volume_Change']=='Decrease']

    fig = go.Figure()
    fig.add_trace(go.Bar(x=vdf_increase['Date'], y=vdf_increase['Volume'],
                    base=0,
                    marker_color='green',
                    name='Volume Increase'))
    fig.add_trace(go.Bar(x=vdf_decrease['Date'], y=vdf_decrease['Volume'],
                    base=0,
                    marker_color='red',
                    name='Volume Decrease'
                    ))

    fig.show()
